django/db/models/sql/query.py

```diff
@@ -1416,13 +1430,27 @@ def setup_joins(self, names, opts, alias, can_reuse=None, allow_many=True):
         # Then, add the path to the query's joins. Note that we can't trim
         # joins at this stage - we will need the information about join type
         # of the trimmed joins.
-        for join in path:
+        for pos, join in enumerate(path):
             opts = join.to_opts
             if join.direct:
                 nullable = self.is_nullable(join.join_field)
             else:
                 nullable = True
-            connection = Join(opts.db_table, alias, None, INNER, join.join_field, nullable)
+
+            try:
+                filtered_relation = self._filtered_relations[names[pos]]
+            except (KeyError, IndexError):
+                connection = Join(opts.db_table, alias, None, INNER, join.join_field, nullable)
+            else:
+                table_alias = filtered_relation.alias
+                self.table_map[opts.db_table] = [table_alias]
```

Comment:
Dealing with `self.table_map` , `self.alias_refcount` and `self.alias_map` here, looks dirty.